### PR TITLE
[Fix #13854 (Old issue)] add push/pop logic for handle local enable/disable

### LINKED ISSUE #13854: disable/enable is too imperative
## Is your feature request related to a problem? Please describe.

To disable locally cops, the proposed solution is a pair of `rubocop:disable MyCop/rubocop:enable MyCop`.  But that's really like setting globals, it is not local, it has far reaching consequences.  For instance if you change the global settings in .rubocop.yml and disable MyCop, then `rubocop:enable MyCop` activates it.

It is also quite annoying when using code-generation.  If the generated piece of code include some `rubocop:disable` for some reason, then at the end of it, it might enable something that was disabled on a higher level.

Finally, there's too much redundancy in `rubocop:disable MyCop/rubocop:enable MyCop`.  This forces to edit two lines when, for instance, a new Metrics threshold is reached.  One should suffice.

The global-variable behavior is not very composable.

## Describe the solution you'd like

Something stack-based would be much more convenient.  Say

```ruby
# rubocop:push-disable Cop1, Cop2
[...]
# rubocop:pop
```

I don't know if some people locally enable in addition to disabling, but maybe it would make sense to have

```ruby
# rubocop:push
# rubocop:disable Cop1, Cop2
# rubocop:enable Cop3, Cop4
[...]
# rubocop:pop
```

and `rubocop:push-disable` would simply be syntactic sugar.

Or maybe something terser:

```ruby
# rubocop:push -Cop1, -Cop2, +Cop3, +Cop4
[...]
# rubocop:pop
```

## Describe alternatives you've considered

There are certainly many alternatives for syntax, but I do believe the most important part is to get something composable, and I can't think of anything else but something stack-based.

Cheers!

### PR DESCRIPTION
**Pull request by Magikdidi24**
## Summary

This MR introduces support for **push/pop directives** in RuboCop, providing a cleaner and more intuitive way to manage cop state changes in Ruby code. The push/pop mechanism allows developers to temporarily save the current state of cops, make changes, and restore the previous state, similar to stack operations in programming.

## New Directive Syntax

Two new directive modes are now available:

- `# rubocop:push +Style/Foo -Style/Bar` - Saves the current state of all cops onto a stack (to be restored later)
- `# rubocop:pop` - Restores the most recently saved state from the stack

**Note**: Unlike traditional disable/enable directives, push and pop operate on the entire cop state as a snapshot. pop
  does not accept cop names.

## Key Features

- **Stack-based state management**: Multiple push directives can be nested, with each pop restoring the corresponding push state
- **Automatic state restoration**: No need to manually track which cops were enabled/disabled before making changes
- **Cleaner syntax**: Replaces verbose disable/enable pairs with more semantic push/pop operations
- **Preserves all cop states**: Captures the complete state of all cops at the time of push

### Example 1: Basic Push/Pop - Temporarily Disable a Cop
```ruby
def process_data(input)
  result = input.upcase
  # rubocop:push -Style/GuardClause
  if result.present?
    return result.strip
  end
  # rubocop:pop
  nil
end
```
### Example 2: Temporarily Enable a Disabled Cop
```ruby
When a cop is already disabled globally, you can temporarily re-enable it:
# rubocop:disable Metrics/MethodLength
def long_method
  line1
  line2
  # rubocop:push +Metrics/MethodLength
  def short_method  # This will be checked by MethodLength
    line3
  end
  # rubocop:pop
  line4  # MethodLength disabled again
  line5
end
```
### Example 3: Multiple Cops with Selective Enable
```ruby
When multiple cops are disabled, you can enable just one temporarily:
# rubocop:disable Style/For, Style/Not
for x in [1, 2, 3]
  not x.nil?
end
# rubocop:push +Style/For -Style/GuardClause
for y in [4, 5, 6]    # Style/For checked here
  not y.nil?          # Style/Not still disabled
  if true             # GuardClause disabled here
    return 1
  end
end
# rubocop:pop
for z in [7, 8, 9]    # Both original states restored
  not z.nil?
end
```
### Example 4: Disable New Cop in Push/Pop, Original Stays Enabled
```ruby
Perfect for temporarily disabling a cop for a specific block:
def process
  x = 1
  y = 2
  # rubocop:push -Style/NumericPredicate +Metrics/MethodLength
  if x.size == 0      # NumericPredicate disabled
    puts 'empty'
  end
  if y.size == 1      # NumericPredicate disabled
    puts 'one'
  end
  # rubocop:pop
  if y.size == 2      # NumericPredicate enabled again
    puts 'two'
  end
end
```

### Example 5: Temporary Disable for Problematic Code Block
```ruby
Perfect for legacy code refactoring or complex multi-line assignments:
def configure(stage)
  # rubocop:push -Layout/SpaceAroundMethodCallOperator
  self.stage =
    if    'macro'  .start_with? stage; X::MACRO
    elsif 'dynamic'.start_with? stage; X::DYNAMIC
    elsif 'static' .start_with? stage; X::STATIC
    else raise ArgumentError, "invalid stage: #{stage}"
    end
  # rubocop:pop
  validate_stage  # Original cop state restored
end
```
Related Issue #13854 

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [ ] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by akimd** [comment]
This would definitely solve all the issues I face in my uses of RuboCop with generated code.

And make my hand-written code more natural and easier to maintain: editing two places when I disable a cop locally has always been weird to my eyes.

**Comment by akimd** [comment]
@bbatsov @koic Hi guys,

Is there any issue about this PR?  A stack-based handling of the enable/disable directives is quite natural, and allows to reduce the noise when more local exceptions are needed (currently each exception incurs two lines of diff).  This is exactly what happens in other tools, such as [gcc](https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html) and then [clang](https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-via-pragmas).

It could even be more concise by allowing arguments to `push` (e.g., `rubocop:push +Style/Foo -Style/Bar`).

**Comment by bbatsov** [comment]
I'll take a closer look soon - in general I like the idea, but I'm always concerned when doing wider-impact changes. @rubocop/rubocop-core please, take a look as well when you can, as more feedback here would be welcome. 

> It could even be more concise by allowing arguments to push (e.g., rubocop:push +Style/Foo -Style/Bar).

Yeah, that looks a bit nicer indeed. I'm ok with the current proposal, but if @Magikdidi24 wants to push it further that'd be even better.

Btw, this change will definitely need to be reflected in the user docs, but I guess we can do this only after we're aligned on the implementation.

**Comment by dvandersluis** [comment]
A few general thoughts and edge cases to consider/test for:

* I prefer @akimd's concise example to the existing more verbose style. I also like this more than the verbosity needed currently to explicitly re-enable inline-disabled cops.
* What happens if `rubocop:push` is called multiple times in a row without a corresponding enable/disable directive?
* What happens if `rubocop:push` is called without a corresponding `rubocop:pop`? What about `pop` without a preceding `push`?
* There should be coverage for using this with `rubocop:todo` as well.

**Comment by Magikdidi24** [comment]
I'll handle it, thanks for your feedback.

**Comment by Magikdidi24** [comment]
“Hi @bbatsov and @koic, I’ve written the solution. Could someone give me some feedback on it?”

File: spec/rubocop/comment_config_spec.rb [hunk 1]
```diff
@@ -265,4 +265,333 @@ def some_method
...
+
+  describe 'push/pop directives' do
+    def disabled_lines_of_cop(cop)
+      (1..source.size).each_with_object([]) do |line_number, disabled_lines|
+        enabled = comment_config.cop_enabled_at_line?(cop, line_number)
+        disabled_lines << line_number unless enabled
+      end
+    end
+
+    context 'realistic: temporarily disable a cop for a problematic block' do
```
Comment (bbatsov) [inline spec/rubocop/comment_config_spec.rb]: Why the "realistic" prefix in some of the contexts?

**Comment by bbatsov** [comment]
The code looks OK to me overall. I guess we can proceed to documenting the new functionality before releasing it.

**Comment by bquorning** [comment]
> Two new directive modes are now available:
> 
>     * `# rubocop:push +Style/Foo -Style/Bar` - Saves the current state of all cops onto a stack (to be restored later)
>     * `# rubocop:pop` - Restores the most recently saved state from the stack
> 
> **Note**: Unlike traditional `disable`/`enable` directives, push and pop **do not** accept cop names. They operate on the entire cop state as a snapshot.

This part of the PR description doesn’t seem to be accurate.

> `push` and pop **do not** accept cop names.

The example right above shows `push` accepting two cop names (with `+`/`-` prefix).

Also, there should be a mention of the `# rubocop:push` with no arguments – what does it do?

**Comment by Magikdidi24** [comment]
> > Two new directive modes are now available:
> > ```
> > * `# rubocop:push +Style/Foo -Style/Bar` - Saves the current state of all cops onto a stack (to be restored later)
> > * `# rubocop:pop` - Restores the most recently saved state from the stack
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **Note**: Unlike traditional `disable`/`enable` directives, push and pop **do not** accept cop names. They operate on the entire cop state as a snapshot.
> 
> This part of the PR description doesn’t seem to be accurate.
> 
> > `push` and pop **do not** accept cop names.
> 
> The example right above shows `push` accepting two cop names (with `+`/`-` prefix).
> 
> Also, there should be a mention of the `# rubocop:push` with no arguments – what does it do?

Yes, I copy-pasted it and didn’t look at it — that was a mistake.

If you run rubocop:push without any arguments, it saves the current cops in a snapshot, but because nothing has changed, the new snapshot is identical.

**Comment by bbatsov** [comment]
Thanks for driving this to the finish line!
